### PR TITLE
KAFKA-6688. The Trogdor coordinator should track task statuses

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

**Comment by cmccabe** [comment]
Test failure is `kafka.api.ConsumerBounceTest.testCloseDuringRebalance`, not related

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (apovzner) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java:194]: do we need this `if`, since we are not doing any calculations for arguments we pass to `log.trace`?

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (apovzner) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java]: what's " WATEREMLON"?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java
Comment (apovzner) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java]: Would be great to make it a bit more robust w.r.t. adding/removing percentiles form Histogram, i.e.  But I see we already do this in other places. I can add JIRA to improve that?

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java:194]: `agentStatus#toString` is pretty heavy.  If you have a dozen workers going on, it will serialize information about all of them into a string.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java]: Sorry, I forgot to take that out.  Let me fix up these log messages

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java]: Hmm, Histogram is internally synchronized, so there should be no conflict.  Maybe I misunderstood the question

Comment (apovzner) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java]: Oh I meant that we explicitly get first three values from percentiles

**Comment by cmccabe** [comment]
retest this please

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (rajinisivaram) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java]: @cmccabe Looks like the log entry hasn't been updated.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (rajinisivaram) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java]: Same as before, update log line?

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java
Comment (rajinisivaram) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Missing `nodeName` in the log line?

**Review by rajinisivaram — COMMENTED** [review]
@cmccabe Thanks for the PR. Looks good. Left a few minor comments. There is also an outstanding question from @apovzner about the histogram. Once they are addressed, I can merge this.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java
Comment (apovzner) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java:518]: It looks like before moving this code into a separate method, we handled PENDING ManagedTaskState. Or this state is not possible when we are in this method? Maybe we should check and throw an exception in that case? Also, what if the task is in STOPPING state?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java
Comment (apovzner) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java]: Actually, it's fine, but I think we should put a comment in StatusUpdater constructor that changing percentiles also need to make sure that it's consistent with json properties in StatusData and how we construct it.  The reason is that I just noticed that we actually create these percentiles: `this.percentiles = new float[] {0.50f, 0.95f, 0.99f};` but StatusData has `@JsonProperty("p90LatencyMs") int p90latencyMs` instead of p95.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java]: fixed.  I wasn't able to search for it because it was misspelled :disappointed:

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java:518]: > It looks like before moving this code into a separate method, we handled PENDING ManagedTaskState. Or this state is not possible when we are in this method?

Yes, PENDING is impossible here because at that point the worker hasn't been started.

> Also, what if the task is in STOPPING state?

There's no additional action needed in that case.

The only transition here is that if one worker fails while a task is RUNNING, the task will transition into STOPPING and the other workers will be stopped.  There is nothing to do if the task isn't RUNNING.

Actually, come to think of it, we should probably not start stopping the other tasks unless the first worker stopped with an error.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java
Comment (rajinisivaram) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: We want `task.error` in the log entry?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java]: Good catch, @apovzner .  I fixed the discrepancy.  I also moved the array into the StatusUpdater class and added a comment.

**Comment by apovzner** [comment]
Looks like my changes conflicted with yours, but there is also a unit test failure (before the conflict happened):
```
17:02:00 org.apache.kafka.trogdor.agent.AgentTest > testWorkerCompletions FAILED
17:02:00     java.lang.AssertionError: Condition not met within timeout 15000. Timed out waiting for expected workers {"bar":{"id":"bar","workerState":{"state":"RUNNING","spec":{"class":"org.apache.kafka.trogdor.task.SampleTaskSpec","startMs":0,"durationMs":900000,"nodeToExitMs":{"node01":2},"error":"baz"},"startedMs":0,"status":"started"}},"foo":{"id":"foo","workerState":{"state":"DONE","spec":{"class":"org.apache.kafka.trogdor.task.SampleTaskSpec","startMs":0,"durationMs":900000,"nodeToExitMs":{"node01":1}},"startedMs":0,"doneMs":1,"status":"halted"}}}
17:02:00         at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:276)
17:02:00         at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:254)
17:02:00         at org.apache.kafka.trogdor.common.ExpectedTasks.waitFor(ExpectedTasks.java:176)
17:02:00         at org.apache.kafka.trogdor.agent.AgentTest.testWorkerCompletions(AgentTest.java:239)
17:02:00 
```

**Comment by cmccabe** [comment]
Rebased and fixed failing test

**Review by rajinisivaram — APPROVED** [review]
@cmccabe Thanks for the updates, LGTM. Will merge after builds complete.

**Comment by apovzner** [comment]
LGTM

**Comment by rajinisivaram** [comment]
@cmccabe Thanks for the PR. Build failure is unrelated, merging to trunk.

**Pull request merged**
